include/Geography.hpp

```diff
@@ -0,0 +1,199 @@
+namespace voltdb {
+
+class ByteBuffer;
+
+struct Geography {
+    struct Ring {
+        Ring() {}
+
+        void clear() {
+            m_points.clear();
+        }
+
+        Ring &addPoint(const GeographyPoint &aPoint) {
+            m_points.push_back(aPoint);
+            return *this;
+        }
+
+        int numPoints() const {
+            return m_points.size();
+        }
+
+        const GeographyPoint &getPoint(int idx) const {
+            return m_points[idx];
+        }
+
+        GeographyPoint &getPoint(int idx) {
+            return m_points[idx];
+        }
+
+        /**
+         * Sometimes rings need to be reversed.  We do this by reversing
+         * all but the first and last. The first and last should be equal.
+         */
+        void reverse();
+
+        bool operator==(const Ring &aOther) const {
+            return approximatelyEqual(aOther, 0.0);
+        }
+
+        bool operator!=(const Ring &aOther) const {
+            return !operator==(aOther);
+        }
+
+        bool approximatelyEqual(const Ring &rhs, double epsilon) const;
+
+        std::string toString() const;
+                       
+        void serializeTo(ByteBuffer &buffer, bool reverseit) const;
+    private:
+        typedef std::vector<GeographyPoint> PointVector;
+        typedef PointVector::iterator       PointIterator;
+        typedef PointVector::const_iterator PointConstIterator;
+        PointVector                         m_points;
+    };
+
+    Geography() {}
+
+    /**
+     * Add an already existing ring.  This is used in
+     * operator<< to add rings to Geographies.  A much
+     * more efficient location is to sue addEmptyRing()
+     * (c.f. below).
+     */
+    Geography &addRing(const Ring &aRing) {
+        m_rings.push_back(aRing);
+        return *this;
+    }
+    /**
+     * Add a new ring, then return a reference to the ring we added.
+     * This is used typically like this:
+     *   polygon.addEmptyRing() << GeographyPoint(0, 0)
+     *                          << GeographyPoint(1, 0)
+     *                          ...;
+     */
+    Ring &addEmptyRing() {
+        // It's unfortunate that we can't use C++11
+        // here.  This is perfect for emplacement new.
+        m_rings.push_back(Ring());
+        return m_rings[m_rings.size()-1];
+    }
+
+    int numRings() const {
+        return m_rings.size();
+    }
+
+    const Ring &getRing(int idx) const {
+        return m_rings[idx];
+    }
+
+    bool operator==(const Geography &aOther) const {
+        return approximatelyEqual(aOther, 0.0);
+    }
+
+    bool operator !=(const Geography &aOther) const {
+        return !operator==(aOther);
+    }
+
+    /**
+     * Return true iff the the given geography is equal to this
+     * one, but allow the coordinates to differ by something less
+     * than epsilon.  If epsilon is 0.0, we require strict
+     * equality.
+     */
+    bool approximatelyEqual(const Geography &rhs, double epsilon) const ;
```

Comment:
No default value for epsilon?

Users may be confused about what to put here and why this parameter is necessary.  Since we have to convert between lng/lat and (x,y,z) and back, there is an inherent epsilon based on this round trip.  In Java don't we use something like 10^-9 or 10^-12?